haystack/schema.py

```diff
@@ -549,23 +566,24 @@ def no_answer(self) -> Optional[bool]:
         return no_answer
 
     def to_dict(self):
-        return asdict(self)
+        return asdict(self, dict_factory=_dict_factory)
 
     @classmethod
-    def from_dict(cls, dict: dict):
-        # backward compatibility for old labels using answers with document_id instead of document_ids
+    def from_dict(cls, dict: "dict"):
         answer = dict.get("answer")
-        if answer and "document_id" in answer:
-            dict = dict.copy()
+        if answer and isinstance(answer, builtins.dict):
             dict["answer"] = Answer.from_dict(dict["answer"])
-        return _pydantic_dataclass_from_dict(dict=dict, pydantic_dataclass_type=cls)
+        doc = dict.get("document")
+        if doc and isinstance(doc, builtins.dict):
+            dict["document"] = Document.from_dict(dict["document"])
```

Comment:
I'm not a huge fan of the walrus operator since I find it difficult to read. So if its okay with you I'll just remove the unnecessary `if doc` but leave the rest.

Reply:
Good for me, it's true that it might not be easy to read some times.

Reply:
Sorry that I missed the `isinstance(doc, Dict)` here.